HDDS-10374. Make container scanner generate merkle trees during the scan

errose28 opened the pull request:
## What changes were proposed in this pull request?

Allow the container scanner to update the container Merkle Tree with the currently calculated hashes of what is on disk. If the tree does not already exist, it will be created. Even if the container is unhealthy, the merkle tree should still be updated to reflect the current state of the container. Hashes stored in RocksDB that were given by the client during write will not be modified.

Specific changes:
- `KeyValueContainerCheck` has been modified to support building the merkle tree from the data the scanner sees.
- Reworked the container checksum update methods in `KeyValueHandler`.
  - In particular added a public `updateContainerChecksum` method for scanners to call so that ICRs are sent for checksum updates even if the container's state does not change.
  - This also allows the `KeyValueHandler` to manage the in-memory data checksum update in the `ContainerData` object, so we don't need to do it from within `ContainerChecksumTreeManager` anymore.
- A new `ContainerScanHelper` class was created to consolidate duplicate code among the scanners.
- `ContainerMerkleTree` can now be built to contain blocks without any chunks, which is necessary to represent states like an empty block file.

## What is the link to the Apache JIRA

HDDS-10374

## How was this patch tested?

This patch contains unit, integration, and acceptance test additions and modifications. Major ones to note:
- Unit tests for all container scanners updated to test that failure to update checksum does not impact moving a container to unhealthy.
- The new `KeyValueHandler#updateContainerChecksum` method has a unit test added to check both ICR and disk updates happen correctly.
- The `TestContainerReconciliationWithMockDatanodes` suite has been enabled after #8402
- Container scanner integration tests have been updated to check container checksums after corruption is injected.
- Existing tests have been migrated to use the on-demand scanner to generate merkle trees instead of manually building them. This removes numerous TODOs from the code.

More robust testing of fault combinations will be added later using the framework being developed under HDDS-11942

aswinshakil commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueContainerCheck.java:
```diff
@@ -379,12 +378,13 @@ private List<ContainerScanError> scanBlock(DBHandle db, File dbFile, BlockData b
             blockErrors.add(error);
           }
         } else if (chunk.getChecksumData().getType() != ContainerProtos.ChecksumType.NONE) {
+          currentTree.addBlock(block.getBlockID().getLocalID());
```
Can we not add the block when we add the first chunk?

errose28 replied:
The [truncated block](https://github.com/apache/ozone/blob/a3401a93835e59193feedf0f3528dd9cd398c1a5/hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/keyvalue/TestContainerCorruptions.java#L113) failure is currently testing this path.

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -1348,7 +1358,6 @@ public void markContainerUnhealthy(Container container, ScanResult reason)
     } finally {
       container.writeUnlock();
     }
-    createContainerMerkleTree(container);
```
Yes, when we have an unhealthy container we should build the tree from the actual data, but this method only builds the tree from the metadata. I'm renaming it in the next commit to make that clearer. We should not depend on just the metadata if we suspect corruption. The way this should work is open -> unhealthy would trigger an on-demand scan of the container, but that code path is not in place right now. We may want to add that to our branch actually.

aswinshakil replied:
Yes, I think we should add the on-demand scan here to generate the Merkle tree.

aswinshakil reviewed (commented):
Overall the change looks good to me. I have left a minor comment to my previous comment.

aswinshakil commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/checksum/ContainerChecksumTreeManager.java:
```diff
@@ -99,12 +104,18 @@ public void writeContainerDataTree(ContainerData data, ContainerMerkleTree tree)
+      LOG.debug("Data merkle tree for container {} updated with container checksum {}", containerID, dataChecksum);
     } finally {
+      // Even if persisting the tree fails, we should still update the data checksum in memory to report back to SCM.
+      data.setDataChecksum(dataChecksum);
```
Whenever we update the checksum for any reason we need to send an ICR to SCM.

errose28 replied:
This is handled in the latest commits.

aswinshakil commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/BackgroundContainerDataScanner.java:
```diff
@@ -92,6 +92,13 @@ public void scanContainer(Container<?> c)
+      try {
+        checksumManager.writeContainerDataTree(containerData, result.getDataTree());
+      } catch (IOException ex) {
+        LOG.error("Failed to write container merkle tree for container {}", containerId, ex);
```
Same here, need to send an ICR if the checksum changes.

errose28 replied:
This is handled in the latest commits.

hemantk-12 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerScanError.java:
```diff
@@ -32,9 +32,10 @@ public enum FailureType {
     MISSING_CONTAINER_FILE,
     MISSING_CHUNKS_DIR,
-    MISSING_CHUNK_FILE,
+    MISSING_DATA_FILE,
```
qq: Is this for the cases when `*.block` file is missing?

errose28 replied:
Yes. I changed the name so it is agnostic of container layout.

hemantk-12 reviewed (commented):
Thanks @errose28 for the patch.

aswinshakil reviewed (commented):
One thing that struck my mind is that it's not related to this PR, but we can have it as a separate JIRA. We shouldn't allow reconciliation of a container that is already reconciling. This can lead to some other bugs where we update the metadata or the data in parallel.

errose28 commented:
> We shouldn't allow reconciliation of a container that is already reconciling. This can lead to some other bugs where we update the metadata or the data in parallel.

Most of this should be handled for us by the [ReplicationSupervisor](https://github.com/apache/ozone/blob/1aef378d70f11ed7523c397903b3a26edf93c635/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/replication/ReplicationSupervisor.java#L98) and tested generally [here](https://github.com/apache/ozone/blob/2b4708b70e5c9de133381a38ca7fa4b3cf3caa42/hadoop-hdds/container-service/src/test/java/org/apache/hadoop/ozone/container/replication/TestReplicationSupervisor.java#L558). However this depends on how we define the [uniqueness of the reconciliation commands](https://github.com/apache/ozone/blob/989746a0de358267be31ce81824046766b49bb7c/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/protocol/commands/ReconcileContainerCommand.java#L98) and here I think we do have a problem as you mentioned, because we can get multiple simultaneous reconciliation tasks for the same container if the peer list is different. We should fix this in a follow-up so that `ReconcileContainerCommand` equality is based only on container ID, and add a test for this that is unique to reconciliation commands.

errose28 commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -1394,11 +1368,97 @@ public void markContainerForClose(Container container)
   }
 
+  @Override
+  public void updateContainerChecksum(Container container, ContainerMerkleTreeWriter treeWriter)
```
I'm adding this to the javadoc in the interface. None of the other `Handler` methods that send an ICR have `withICR` or similar in their name. It seems this is supposed to be more of an internal detail.

errose28 commented:
Thanks for the review @aswinshakil. All comments should be addressed. I still need to stabilize [this test](https://github.com/apache/ozone/actions/runs/15072765850/job/42401912162) on our branch before trying to get green CI with this change.

aswinshakil commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/keyvalue/KeyValueHandler.java:
```diff
@@ -1427,7 +1487,7 @@ public void markContainerUnhealthy(Container container, ScanResult reason)
       container.writeUnlock();
     }
-    createContainerMerkleTreeFromMetadata(container);
+    updateContainerChecksumFromMetadataIfNeeded(container);
```
https://github.com/apache/ozone/pull/8402#discussion_r2080078695 In #8402 we were planning to do an On-demand scanning here rather than generating the merkle tree from the metadata.

errose28 replied:
I filed HDDS-13084 to do this in a follow-up. It will need testing that would expand this already large patch.

aswinshakil commented on hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/ozoneimpl/ContainerScanHelper.java:
```diff
@@ -0,0 +1,154 @@
+    // never scanned.
+    boolean recentlyScanned = lastScanTime.map(scanInstant ->
+            Duration.between(now, scanInstant).abs()
+                .compareTo(Duration.ofMillis(minScanGap)) < 0)
```
This is specific to `minScanGap` for On-demand scanner. Right now, the default is 15m. Which means we have to wait for the on-demand scanner to run on the unhealthy container and wait 15 mins before triggering reconciliation as reconciliation also triggers an on-demand scan at the end, That might get dropped because of this.

aswinshakil commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/checksum/TestContainerCommandReconciliation.java:
```diff
@@ -557,22 +515,15 @@ public void testDataChecksumReportedAtSCM() throws Exception {
-    // TODO: Use On-demand container scanner to build the new container merkle tree. (HDDS-10374)
-    Files.deleteIfExists(getContainerChecksumFile(container.getContainerData()).toPath());
-    kvHandler.createContainerMerkleTreeFromMetadata(container);
+    datanodeStateMachine.getContainer().getContainerSet().scanContainer(containerID);
```
```suggestion
   datanodeStateMachine.getContainer().getContainerSet().scanContainerWithoutGap(containerID);
```
Need to change this in the whole test suite.

aswinshakil commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/checksum/TestContainerCommandReconciliation.java:
```diff
@@ -460,42 +445,15 @@ public void testContainerChecksumChunkCorruption() throws Exception {
-    Files.deleteIfExists(getContainerChecksumFile(container.getContainerData()).toPath());
-    kvHandler.createContainerMerkleTreeFromMetadata(container);
-    // To set unhealthy for chunks that are corrupted.
+    datanodeStateMachine.getContainer().getContainerSet().scanContainer(containerID);
```
Same here

aswinshakil commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/checksum/TestContainerCommandReconciliation.java:
```diff
@@ -376,21 +368,14 @@ public void testContainerChecksumWithBlockMissing() throws Exception {
-    // TODO: Use On-demand container scanner to build the new container merkle tree. (HDDS-10374)
-    Files.deleteIfExists(getContainerChecksumFile(container.getContainerData()).toPath());
-    kvHandler.createContainerMerkleTreeFromMetadata(container);
+    datanodeStateMachine.getContainer().getContainerSet().scanContainer(containerID);
```
same here

aswinshakil reviewed (commented):
`TestContainerCommandReconciliation#testContainerChecksumChunkCorruption` seems to be failing, I tried to look into it. There seems to be an issue with the scan container logic. Can you take a look at it?

aswinshakil commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/dn/checksum/TestContainerCommandReconciliation.java:
```diff
@@ -148,10 +144,6 @@ public static void init() throws Exception {
     conf.set(OZONE_METADATA_DIRS, testDir.getAbsolutePath());
     conf.setStorageSize(OZONE_SCM_CHUNK_SIZE_KEY, 128 * 1024, StorageUnit.BYTES);
```
```suggestion
    conf.set(OZONE_SCM_HEARTBEAT_PROCESS_INTERVAL, "200ms");
    conf.set(HDDS_HEARTBEAT_INTERVAL, "1s");
    conf.set(OZONE_SCM_STALENODE_INTERVAL, "3s");
    conf.set(OZONE_SCM_DEADNODE_INTERVAL, "6s");
    conf.set(HDDS_NODE_REPORT_INTERVAL, "5s");
    conf.set(HDDS_CONTAINER_REPORT_INTERVAL, "5s");
    conf.setStorageSize(OZONE_SCM_CHUNK_SIZE_KEY, 128 * 1024, StorageUnit.BYTES);
```
I have tested `testDataChecksumReportedAtSCM` flakiness by adding the following configs and it works, The main reason for this error was that SCM was in safe mode as it didn't receive the container report from the datanode. This fixes it. 
Flaky run details - 100 runs: https://github.com/aswinshakil/ozone/actions/runs/15295475218

errose28 commented:
@aswinshakil I added the test fixes you suggested for the flaky tests on the branch to this PR:
- `TestFilePerBlockStrategy` running out of space
- `TestContainerCommandReconciliation#testDataChecksumReportedAtSCM ` timing out when SCM was restarted

I also fixed `TestContainerCommandReconciliation#testContainerChecksumChunkCorruption` which started failing in this PR. The test was using a non-standard way of introducing corruption by changing the stored client checksums of the replica, not the data, which is something we can't recover from. I changed it to corrupt the actual data instead.

errose28 commented:
Thanks for the reviews @aswinshakil @kerneltime
